### PR TITLE
Added RegisteredLookupDimension

### PR DESCRIPTION
**Pull request by garyluoex**
Only serialization test is included, since the only difference between `RegisteredLookupDimension` and `LookupDimension` is the extraction function serialization. Let me know if more End To End tests should be added in this PR also, thanks.

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    private final String lookup;
+    private final Boolean retainMissingValue;
+    private final String replaceMissingValueWith;
+    private final Boolean injective;
+    private final Boolean optimize;
+
+    /**
+     * Constructor.
+     *
+     * @param lookup  lookup property specified by the user.
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Can we remove the period at the end?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+     * @param retainMissingValue  when true, returns original dimension value if mapping is not found and
+     * replaceMissingValueWith must be null or empty string, when false, missing values are treated as missing.
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Can we remove the period at the end?

Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Can this be indented from the `@param` seems too cluttered.

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    private final Boolean retainMissingValue;
+    private final String replaceMissingValueWith;
+    private final Boolean injective;
+    private final Boolean optimize;
+
+    /**
+     * Constructor.
+     *
+     * @param lookup  lookup property specified by the user.
+     * @param retainMissingValue  when true, returns original dimension value if mapping is not found and
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: The sentence seems ambiguous to understand, can we re-fragment it?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        this.replaceMissingValueWith = replaceMissingValueWith;
+        this.injective = injective;
+        this.optimize = optimize;
+    }
+
+    /**
+     * Convenience Constructor,
+     * defaults: retainMissingValue=false, replaceMissingValueWith=null, injective=false, optimize=true.
+     *
+     * @param lookup  lookup property specified by the user.
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Can we remove the period at the end?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    // CHECKSTYLE:OFF
+    public RegisteredLookupExtractionFunction withLookup(String lookup) {
+        return new RegisteredLookupExtractionFunction(lookup, retainMissingValue, replaceMissingValueWith, injective, optimize);
+    }
+
+    /**
+     * Builds a new LookupExtractionFunction with the given retainMissingValue property, preserves other properties.
+     *
+     * @param retainMissingValue  when true, returns original dimension value if mapping is not found and
+     * replaceMissingValueWith must be null or empty string, when false, missing values are treated as missing.
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Can we remove the period at the end?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    public RegisteredLookupExtractionFunction withRetainMissingValue(Boolean retainMissingValue) {
+        return new RegisteredLookupExtractionFunction(lookup, retainMissingValue, replaceMissingValueWith, injective, optimize);
+    }
+
+    /**
+     * Builds a new LookupExtractionFunction with the given replaceMissingValueWith property,preserves other properties.
+     *
+     * @param replaceMissingValueWith  replaces dimension values not found in mapping with this value and
+     * retainMissingValue must be false if this value is not null or is not empty string.
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Can we remove the period at the end?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+     * replaceMissingValueWith must be null or empty string, when false, missing values are treated as missing.
+     *
+     * @return a new instance of LookupExtractionFunction with the given retainMissingValue
+     */
+    public RegisteredLookupExtractionFunction withRetainMissingValue(Boolean retainMissingValue) {
+        return new RegisteredLookupExtractionFunction(lookup, retainMissingValue, replaceMissingValueWith, injective, optimize);
+    }
+
+    /**
+     * Builds a new LookupExtractionFunction with the given replaceMissingValueWith property,preserves other properties.
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: space after the `,`

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+     *
+     * @param replaceMissingValueWith  replaces dimension values not found in mapping with this value and
+     * retainMissingValue must be false if this value is not null or is not empty string.
+     *
+     * @return a new instance of LookupExtractionFunction with the given replaceMissingValueWith
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: The sentence seems incomplete can we add `string` to the end of the sentence?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+     */
+    public RegisteredLookupExtractionFunction withReplaceMissingValueWith(String replaceMissingValueWith) {
+        return new RegisteredLookupExtractionFunction(lookup, retainMissingValue, replaceMissingValueWith, injective, optimize);
+    }
+
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Why are no `javadocs` added here?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    public RegisteredLookupExtractionFunction withInjective(Boolean injective) {
+        return new RegisteredLookupExtractionFunction(lookup, retainMissingValue, replaceMissingValueWith, injective, optimize);
+    }
+
```
Comment (RVRSKumar) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: Why are no `javadocs` added here?

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/serializers/RegisteredLookupDimensionToDimensionSpecSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2016 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.druid.serializers
```
Comment (RVRSKumar) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/druid/serializers/RegisteredLookupDimensionToDimensionSpecSpec.groovy]: There seems to be a typo in the file name, there are 2 `Spec`. Can we rename the file?

File: fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java (hunk 1 above)
Comment (garyluoex) [inline fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/dimension/extractionfunction/RegisteredLookupExtractionFunction.java]: I wanted to fix the period at the end for a long time already =)

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/ErrorMessageFormat.java [hunk 12]
```diff
@@ -22,6 +22,7 @@
     DIMENSIONS_UNDEFINED("Dimension(s) '%s' do not exist."),
     DIMENSIONS_NOT_IN_TABLE("Requested dimension(s) '%s' are not supported by the table '%s'."),
     DIMENSION_FIELDS_UNDEFINED("Dimension field(s) '%s' do not exist for dimension '%s'"),
+    DIMENSION_TYPE_INVALID("The dimension type %s for dimension %s is not invalid"),
```
Comment (cdeszaq) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/ErrorMessageFormat.java]: Double negative in this message makes this not an error message.

Comment (garyluoex) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/ErrorMessageFormat.java]: ops

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/config/dimension/TypeAwareDimensionLoader.java [hunk 13]
```diff
@@ -35,20 +38,22 @@ public TypeAwareDimensionLoader(Iterable<DimensionConfig> configSource) {
...
+            } else if (dimensionConfig.getType().equals(LookupDimension.class)) {
+                dimensions.add(new LookupDimension((LookupDimensionConfig) dimensionConfig));
+            } else if (dimensionConfig.getType().equals(RegisteredLookupDimension.class)) {
+                dimensions.add(new RegisteredLookupDimension((RegisteredLookupDimensionConfig) dimensionConfig));
             } else {
-                LOG.warn(
-                    String.format(
-                        "The dimension type for the dimension %s is not defined",
-                        dimension.getApiName())
+                throw new RuntimeException(
```
Comment (garyluoex) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/config/dimension/TypeAwareDimensionLoader.java]: Ignoring you

**Pull request merged**